Ask for consent - Matomo (Piwik)

jayhaluska opened the pull request:
This is a pull request to add the "Ask for consent" functionality from Matomo (Piwik). Especially relevant in the context of the new GDPR in Europe.

See: https://developer.matomo.org/guides/tracking-javascript-guide#asking-for-consent

jayhaluska committed:
Asking for consent implementation + documentation

bittner commented:
Fixes #141

jayhaluska committed:
Fixed ask for consent test name

bittner reviewed (changes requested):
The changes sound good in general. If we could make it more elegant (e.g. just a single JavaScript call to pull the feature in from Matomo) that would be even more awesome! Do you see a way?

jayhaluska committed:
Removed backslashes from strings, improved documentation text

Paluschki commented on analytical/templatetags/piwik.py:
```diff
@@ -42,6 +42,9 @@
 VARIABLE_CODE = '_paq.push(["setCustomVariable", %(index)s, "%(name)s", "%(value)s", "%(scope)s"]);'  # noqa
 IDENTITY_CODE = '_paq.push(["setUserId", "%(userid)s"]);'
 DISABLE_COOKIES_CODE = '_paq.push([\'disableCookies\']);'
+ASK_FOR_CONSENT_CODE = '_paq.push([\'requireConsent\']);'
```
Changed in e50b247a87aa1ccc0a091daef417a97ea004345a

Paluschki commented on analytical/tests/test_tag_piwik.py:
```diff
@@ -150,3 +150,8 @@ def test_disable_usertrack(self):
+    @override_settings(PIWIK_ASK_FOR_CONSENT=True)
+    def test_ask_for_consent(self):
+        r = PiwikNode().render(Context({}))
+        self.assertTrue("_paq.push([\'requireConsent\']);" in r, r)
```
Removed backslashes in e50b247a87aa1ccc0a091daef417a97ea004345a

Paluschki commented on docs/services/piwik.rst:
```diff
@@ -145,11 +145,24 @@ set the context variable ``analytical_identity`` (for global configuration) or
+-----------------
+
+If you want to ask for consent set :data:`PIWIK_ASK_FOR_CONSENT` to
+:const:`True`. This is disabled by default.
```
Changed documentation text in e50b247a87aa1ccc0a091daef417a97ea004345a

Paluschki commented on docs/services/piwik.rst:
```diff
@@ -145,11 +145,24 @@ set the context variable ``analytical_identity`` (for global configuration) or
+To ask the visitor for consent just create DOM elements with the following id's:
+
+`piwik_deny_consent` - id for DOM element to click, if the user denies consent
+`piwik_give_consent` - id for DOM element to click, if the user gives consent
```
Yes, Matomo provides an opt-out form. But as far as I understood the concept of "Ask for consent", it is quite independent to the idea of the opt-out form.

With the opt-out form you are tracked until you opt-out (obviously).
With asking for consent, you are not tracked until you deliberately accept the tracking. Therefore, a user/visitor interaction is needed. 

In my opinion, this is the easiest way to implement that feature. The user of django-analytical just has to insert two buttons to his consent request box with these id's.

bittner replied:
What I'm worried about is that this may be already "too difficult" for a lot of average Django users. You know, developers are spoiled, they want to plug an app in and see it working. :see_no_evil: :grimacing: 

Seriously, the buttons are not the problem, the styling and overall user experience is. ("How should this thing of 'consent or no' look like?", "Where do I place the buttons?", "Do I make them disappear once they are clicked?", etc.)

Don't you think this is challenging for some developers? Or they'd come up with mediocre results (and people start re-iterating the "Wordpress is much better" :smirk:, because Wordpress plugin are designed to have a ready-to-use plug-and-play developer experience).

Can we solve this issue? Matomo doesn't provide anything we can just pull in, does it? Or do the Matomo docs provide a UX design example? (I didn't find any.)

Paluschki replied:
I can understand your concern regarding usability. Nevertheless, I don't want to constrain the flexibility to implement the consent buttons as the user wants. I thought one of the reasons users choose Django over Wordpress is that they have more flexibility. Correct me if that's not the case ;)

The idea for a template tag is very good. I would implement it with a start and end tag, like this:

`{ piwik_allow }
<button>Track me!</button>
{ end_piwik_allow }`

You may ask: why not just one tag which uses under the hood a HTML button?
For most of the cases buttons might suffice. But sometimes you want to have only a clickable image, or a font-awesome icon. To preserve flexibility the tag should be able to use arbitrary HTML elements inside of it.

`{ piwik_deny }
<i fa-2x fa-cross></i>
{ end_piwik_deny }`

Is this a possible solution or is it then again too complex?

bittner replied:
I think the latter is a particularly good idea! And, yes, it's certainly possible (take a look at the [built-in template tags](https://docs.djangoproject.com/en/stable/ref/templates/builtins/) for reference).

The [simple_tag](https://docs.djangoproject.com/en/stable/howto/custom-template-tags/#writing-custom-template-tags) can have arguments, so our API may look like this:

```django
{ piwik_allow }  yields just  id="..."  (as in your example)
```
```django
{ piwik_allow button }  yields  <button id="...">Allow</button>  (localized)
```
```django
{ piwik_allow button "Sure!" }  yields  <button id="...">Sure!</button>
```
```django
{ piwik_deny div }  yields  <div id="...">Deny</div>  (just like <button>)
```
```django
{ piwik_deny i "fa-2x fa-cross" }  yields  <i id="..." fa-2x fa-cross></i>
```
Just allow a few ones. Depending on the to-be-generated HTML tag you validate and use arguments 2+ the appropriate way (e.g. use parameter 3 for `class="..."` or so, if that's really needed).

That looks sufficiently clean, doesn't it? How do you like it? :smiley:

Paluschki replied:
I have implemented it like you proposed, but now I have doubts.

I like the idea but it's maybe problematic that sometimes the template tag returns fully fledged HTML and sometimes it returns just an id.

Also, there are many edge cases, like what if a user wants to set the "type" attribute of an `input` element, or the user leaves out some parameters. This means I have to enhance the template tag further, and validate the input, etc.

Furthermore, now the user writes
`{% piwik_remember_consent "button" "Please track me!" "buttonClass" %}`
instead of 
`<button {% piwik_remember_consent %} class="buttonClass">Track me!</button>`

In my opinion, this is even more difficult to use, in case you don't know template tags very well.

I would go with a simple template tag: the user builds his own HTML and attaches the id with the tag. This would keep the template tag function slim and coherent and gives the whole flexibility to the user, while we do not have to worry about edge cases.

What do you think?

bittner replied:
Sure, the usage you prefer would still be the default. Hence, I wouldn't worry.

The "button" & friends, on the other hand, I see as nice idioms: You say, _"Don't give me just _something_, which helps me to construct a consent UI control: Give me a button!"_

Note that I wouldn't use quotes for writing _"button"_, e.g.
```django
{% piwik_remember_consent button %}
```
You have the full freedom with the simple tag and get a nice, of course modified but idiomatic, result with additional parameters.

Let's get another opinion on board, though: @jcassee, what do you think?

bittner replied:
Don't worry too much about edge cases, btw. It's fine when the short idiomatic solutions only cover the principal use cases.

jcassee replied:
First of all: I'm not sure that pushing the event to Piwik is the _only_ thing the consent buttons need to do. They probably also need to close a dialog or fixed banner. Depending on the JS tech used, you need the tag id to integrate with other interactions. This make me think that using classes is better than using the ids.

Looking over the different options, the original one still seems simplest (using a class):
```html
<button class="big green piwik_give_consent">I agree!</button>
```

Using a Django tag that "hides" the HTML tag makes styling with CSS a bit weirder and is not easily customizable without abstracting HTML attributes inside the Django tag (`{ piwik_deny i "fa-2x fa-cross" }`). Wrapping an HTML tag with a Django tag (`{ piwik_deny } <i fa-2x fa-cross></i> { end_piwik_deny }`) looks pretty complex.

Paluschki replied:
I implemented it now as @jcassee proposed because I think it makes most sense in this way and doesn't get to convoluted. Please check the new version and let me know if there is something missing to get this pull request merged.

Thanks for the fast and nice collaboration, I really enjoyed my first PR :)

jayhaluska committed:
Fixed line length (now is < 100)

jcassee commented on analytical/templatetags/piwik.py:
```diff
@@ -41,7 +41,19 @@
+"""
+REMEMBER_CONSENT_CODE = """ document.getElementById("piwik_give_consent").addEventListener("click",
+() => {
+  _paq.push(["setConsentGiven"]); _paq.push(["rememberConsentGiven"]);
```
From the [docs you linked](https://developer.matomo.org/guides/tracking-javascript-guide#step-3-user-gives-consent), it seems only `_paq.push(['rememberConsentGiven']);` is required?

Paluschki replied:
You're right. Changed it in c3065fc0c96e762f064176f16554af440055cadc

jcassee commented on analytical/templatetags/piwik.py:
```diff
@@ -96,6 +108,11 @@ def render(self, context):
+        if getattr(settings, 'PIWIK_ASK_FOR_CONSENT', False):
+            commands.append(ASK_FOR_CONSENT_CODE)
+            commands.append(FORGET_CONSENT_CODE)
+            commands.append(REMEMBER_CONSENT_CODE)
```
I would be okay with folding the contents of these variables into one. Their content is only used together.

Paluschki replied:
Changed in c3065fc0c96e762f064176f16554af440055cadc

jcassee commented on analytical/templatetags/piwik.py:
```diff
@@ -41,7 +41,19 @@
+ASK_FOR_CONSENT_CODE = "_paq.push(['requireConsent']);"
+FORGET_CONSENT_CODE = """
+document.getElementById("piwik_deny_consent").addEventListener("click",
+() => {
```
No real reason not to support poor souls forced to use IE, right? Replace with old-school `function () { ... }`?

Paluschki replied:
Changed in c3065fc0c96e762f064176f16554af440055cadc

jayhaluska committed:
Changed usage of id to class name. Code adjustments (compatibility, naming)

jcassee reviewed (approved):
Looks good to me! Thanks for the work, Julian! @bittner?

jayhaluska committed:
Merge remote-tracking branch 'upstream/master'

jayhaluska commented:
Merged the newest version of the django-analytical into my fork.

I think this PR was forgotten, can we get it merged, @bittner?

bittner commented:
Probably it makes more sense now to add these changes to the newly created Matomo tag.

jayhaluska committed:
Added 'ask for consent' logic and documentation to new matomo files

jayhaluska committed:
Adjusted remaining 'piwik' naming to 'matomo'

jayhaluska commented:
I added the 'ask for consent' logic and documentation to the new matomo files and renamed everything accordingly.

But now the Travis CI check fails and I don't know why. Any suggestions?

bittner commented:
> But now the Travis CI check fails and I don't know why. Any suggestions?

Flake8 complains about a missing newline in the test file:

> ./analytical/tests/test_tag_matomo.py:157:66: W292 no newline at end of file

It's a good idea to configure your IDE in a way that it ensures a final newline in every file you touch. Most modern IDEs have such a setting now.

bittner commented on analytical/templatetags/piwik.py:
```diff
@@ -98,6 +122,9 @@ def render(self, context):
         if getattr(settings, 'PIWIK_DISABLE_COOKIES', False):
             commands.append(DISABLE_COOKIES_CODE)
 
+        if getattr(settings, 'PIWIK_ASK_FOR_CONSENT', False):
```
This still says, `PIWIK` in the constant name. Shouldn't it say `MATOMO`?

bittner replied:
Actually, this code shouldn't go here at all. As mentioned by my other comments.

bittner commented on docs/services/piwik.rst:
```diff
@@ -153,11 +153,31 @@ set the context variable ``analytical_identity`` (for global configuration) or
+    <button class="piwik_remove_consent">Don't track me anymore!</button>
+
+.. _`asking for consent`: https://developer.matomo.org/guides/tracking-javascript-guide#asking-for-consent
+
```
Can't we remove the documentation for Piwiki all together, maybe just showing a hint on the new location and the reason for it? I don't think it makes sense to continue to maintain, let alone extend, it.

bittner commented on analytical/tests/test_tag_matomo.py:
```diff
@@ -150,3 +150,8 @@ def test_disable_usertrack(self):
+    @override_settings(MATOMO_ASK_FOR_CONSENT=True)
+    def test_ask_for_consent(self):
+        r = MatomoNode().render(Context({}))
+        self.assertTrue("_paq.push(['requireConsent']);" in r, r)
```
Here is the missing newline that causes Flake8 to fail.

bittner commented on analytical/templatetags/piwik.py:
```diff
@@ -43,7 +43,31 @@
+      }}
+    );
+}}
+""".format(REMOVE_CONSENT_CLASS, GIVE_CONSENT_CLASS)
```
We should really just keep the Piwik implementation as is. Not extend it.

I'm not even sure if it would work. There is no Piwik that does consent. It's only Matomo that does it.

bittner commented on analytical/tests/test_tag_piwik.py:
```diff
@@ -150,3 +150,8 @@ def test_disable_usertrack(self):
+    @override_settings(PIWIK_ASK_FOR_CONSENT=True)
+    def test_ask_for_consent(self):
+        r = PiwikNode().render(Context({}))
+        self.assertTrue("_paq.push(['requireConsent']);" in r, r)
```
Same here: Why add code and tests for an obsolete integration?

bittner reviewed (commented):
Not sure if we have discussed this before. (Sorry, if we have!)

bittner commented:
This PR needs to be rebased and its conflicts resolved. In addition, some review comments may still need to be addressed. But this it would be ready to be merged.

Want to give it another try?

bittner commented:
@jayhaluska Please rebase this PR and make the pipeline pass, so we can merge your changes. Thank you!

ronardcaktus commented:
Hello @bittner. I recently joined Jazz Band and I'm picking some projects to work on consistently, this one being one of them. I'm happy to continue work on this branch if that's OK with you, unless @jayhaluska wants resume his work.

bittner commented:
Sure, go ahead! Your support is more than welcome! :+1:

bittner commented:
Superseded by #245.

Thank you for your contribution, @jayhaluska!
